nav2_amcl/src/amcl_node.cpp

```diff
@@ -1037,9 +1037,44 @@ AmclNode::publishParticleCloud(const pf_sample_set_t * set)
 }
 
 bool
-AmclNode::getMaxWeightHyp(
-  std::vector<amcl_hyp_t> & hyps, amcl_hyp_t & max_weight_hyps,
-  int & max_weight_hyp)
+AmclNode::getWeightedMeanClustersCenroid(amcl_hyp_t & mean_centroid){
+    double weighted_x = 0.0;    
+    double weighted_y = 0.0;
+    double weighted_yaw = 0.0;
+    double mean_weight = 0.0;
+    for (int cluster_idx = 0; 
+      cluster_idx < pf_->sets[pf_->current_set].cluster_count; cluster_idx++)
+    {
+      double weight;
+      pf_vector_t pose_mean;
+      pf_matrix_t pose_cov;
+      if (!pf_get_cluster_stats(pf_, cluster_idx, &weight, &pose_mean, &pose_cov)) {
+        RCLCPP_ERROR(get_logger(), "Couldn't get stats on cluster %d", cluster_idx);
+        return false;
+      }
+
+      weighted_x += pose_mean.v[0] * weight;
+      weighted_y += pose_mean.v[1] * weight;
+      weighted_yaw += pose_mean.v[2] * weight;
+
+      mean_weight += weight;
+    }
+
+    if(mean_weight == 0.0) return false;
+
+    mean_weight /= pf_->sets[pf_->current_set].cluster_count;
```

Comment:
are the weights nomalised? ie they add up to 1?  If so, then should we not leave `mean_weight` as 1?  And if not, is it correct to multiply x, y and yaw by weight without further normalisation?

Reply:
Yes, all cluster weights add up to 1. And yes, it makes much more sense to leave `mean_weight` equal to 1

Reply:
Remove what averaging? Can you please clarify what you mean?

The weight of the best hypothesis isn't used anywhere, btw

Reply:
After oral discussion, I've got Aaron's idea, implemented in the latest commit.